src/EntityFramework/Core/Query/PlanCompiler/TransformationRulesContext.cs

```diff
@@ -262,6 +262,49 @@ internal bool IsScalarOpTree(Node node, Dictionary<Var, int> varRefMap)
             return IsScalarOpTree(node, varRefMap, ref nodeCount);
         }
 
+        /// <summary>
+        /// Is this tree uses user-defined functions
+        /// Simplifing query with UDFs could caused to suboptimal plans
+        /// </summary>
+        /// <param name="node"> Current subtree to process </param>
+        /// <param name="varMap">Mapped variables</param>
+        /// <returns></returns>
+        internal bool IncludeCustomFunctionOp(Node node, Dictionary<Var, Node> varMap)
+        {
+            PlanCompiler.Assert(varMap != null, "Null varRef map");
+
+            if (node.Op.OpType == OpType.VarRef)
+            {
+                var varRefOp = (VarRefOp)node.Op;
+                if (varMap.TryGetValue(varRefOp.Var, out var newNode))
+                {
+                    if (IncludeCustomFunctionOp(newNode, varMap))
+                    {
+                        return true;
+                    }
+                }
+            }
+            if (node.Op.OpType == OpType.Function)
+            {
+                var functionOp = node.Op as FunctionOp;
+                if (!functionOp.Function.BuiltInAttribute)
+                {
+                    return true;
+                }
+            }
+
+            // Simply process the result of the children.
+            for (var i = 0; i < node.Children.Count; i++)
+            {
+                if (IncludeCustomFunctionOp(node.Children[i], varMap))
```

Comment:
Again, just return the result of the call rather than having the `if`.

Reply:
We could have child without custom function before child with custom function. As a result we need to do Any(…), so I need if there, 

I modified tests to demo that case